common/src/main/java/com/viaversion/viaversion/protocols/protocol1_9to1_8/packets/WorldPackets.java

```diff
@@ -132,22 +137,52 @@ public void registerMap() {
                         ClientChunks clientChunks = wrapper.user().get(ClientChunks.class);
-                        Chunk1_9to1_8Type type = new Chunk1_9to1_8Type(clientChunks);
-                        Chunk1_8 chunk = (Chunk1_8) wrapper.read(type);
-                        if (chunk.isUnloadPacket()) {
-                            wrapper.setPacketType(ClientboundPackets1_9.UNLOAD_CHUNK);
+                        Chunk chunk = wrapper.read(new Chunk1_8Type(clientWorld));
 
+                        // Check if the chunk should be handled as an unload packet
+                        if (chunk.isFullChunk() && chunk.getBitmask() == 0) {
+                            wrapper.setPacketType(ClientboundPackets1_9.UNLOAD_CHUNK);
                             wrapper.write(Type.INT, chunk.getX());
                             wrapper.write(Type.INT, chunk.getZ());
+
                             // Remove commandBlocks on chunk unload
                             CommandBlockProvider provider = Via.getManager().getProviders().get(CommandBlockProvider.class);
                             provider.unloadChunk(wrapper.user(), chunk.getX(), chunk.getZ());
+
+                            clientChunks.getLoadedChunks().remove(ClientChunks.toLong(chunk.getX(), chunk.getZ()));
+
+                            // Unload the empty chunks
+                            if (Via.getConfig().isChunkBorderFix()) {
+                                for (BlockFace face : BlockFace.values()) {
+                                    if (face.getAxis().equals(BlockFace.EnumAxis.Y)) continue;
+                                    if (!clientChunks.getLoadedChunks().contains(ClientChunks.toLong(chunk.getX() + face.getModX(), chunk.getZ() + face.getModZ()))) {
+                                        PacketWrapper unloadChunk = wrapper.create(ClientboundPackets1_9.UNLOAD_CHUNK);
+                                        unloadChunk.write(Type.INT, chunk.getX() + face.getModX());
+                                        unloadChunk.write(Type.INT, chunk.getZ() + face.getModZ());
+                                        unloadChunk.send(Protocol1_9To1_8.class);
+                                    }
+                                }
+                            }
                         } else {
-                            wrapper.write(type, chunk);
-                            // eat any other data (Usually happens with unload packets)
+                            wrapper.write(new Chunk1_9_1_2Type(clientWorld), chunk);
+
+                            clientChunks.getLoadedChunks().add(ClientChunks.toLong(chunk.getX(), chunk.getZ()));
+
+                            // Send empty chunks surrounding the loaded chunk to force 1.9+ clients to render the new chunk
+                            if (Via.getConfig().isChunkBorderFix()) {
+                                for (BlockFace face : BlockFace.values()) {
```

Comment:
Make a private static final array containing the 4 cardinal directions (best put into the BlockFace enum itself), use that in the other places as well.